File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.console.core.entity;

import org.apache.streampark.common.conf.ConfigKeys;
import org.apache.streampark.common.enums.ClusterState;
import org.apache.streampark.common.enums.FlinkDeployMode;
import org.apache.streampark.common.enums.FlinkK8sRestExposedType;
import org.apache.streampark.common.enums.ResolveOrder;
import org.apache.streampark.common.util.HttpClientUtils;
import org.apache.streampark.common.util.PropertiesUtils;
import org.apache.streampark.console.base.util.JacksonUtils;
import org.apache.streampark.console.core.util.YarnQueueLabelExpression;

import org.apache.commons.lang3.StringUtils;
import org.apache.flink.configuration.CoreOptions;
import org.apache.hc.client5.http.config.RequestConfig;

import com.baomidou.mybatisplus.annotation.FieldStrategy;
import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableField;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import lombok.Getter;
import lombok.Setter;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;

import java.io.Serializable;
import java.net.URI;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

@Slf4j
@Getter
@Setter
@TableName("t_flink_cluster")
public class FlinkCluster implements Serializable {

    @TableId(type = IdType.AUTO)
    private Long id;

    @TableField(updateStrategy = FieldStrategy.IGNORED)
    private String address;

    @TableField(updateStrategy = FieldStrategy.IGNORED)
    private String jobManagerUrl;

    private String clusterId;

    private String clusterName;

    private Integer deployMode;

    /** flink version */
    private Long versionId;

    private String k8sNamespace;

    private String serviceAccount;

    private String description;

    private Long userId;

    private String flinkImage;

    private String options;

    private String yarnQueue;

    private Boolean k8sHadoopIntegration;

    private String dynamicProperties;

    private Integer k8sRestExposedType;

    private String k8sConf;

    private Integer resolveOrder;

    @TableField(updateStrategy = FieldStrategy.IGNORED)
    private String exception;

    private Integer clusterState;

    private Date createTime;

    private Date startTime;

    @TableField(updateStrategy = FieldStrategy.IGNORED)
    private Date endTime;

    @TableField(updateStrategy = FieldStrategy.IGNORED)
    private Long alertId;

    private transient Integer allJobs = 0;

    private transient Integer affectedJobs = 0;

    @JsonIgnore
    public FlinkK8sRestExposedType getK8sRestExposedTypeEnum() {
        return FlinkK8sRestExposedType.of(this.k8sRestExposedType);
    }

    @JsonIgnore
    public FlinkDeployMode getFlinkDeployModeEnum() {
        return FlinkDeployMode.of(this.deployMode);
    }

    @JsonIgnore
    public ClusterState getClusterStateEnum() {
        return ClusterState.of(this.clusterState);
    }

    @JsonIgnore
    @SneakyThrows
    public Map<String, Object> getOptionMap() {
        if (StringUtils.isBlank(this.options)) {
            return new HashMap<>();
        }
        Map<String, Object> optionMap = JacksonUtils.read(this.options, Map.class);
        if (FlinkDeployMode.YARN_SESSION == getFlinkDeployModeEnum()) {
            optionMap.put(ConfigKeys.KEY_YARN_APP_NAME(), this.clusterName);
            optionMap.putAll(YarnQueueLabelExpression.getQueueLabelMap(yarnQueue));
        }
        optionMap.entrySet().removeIf(entry -> entry.getValue() == null);
        return optionMap;
    }

    @JsonIgnore
    public URI getRemoteURI() {
        try {
            HttpClientUtils.httpGetRequest(
                this.address,
                RequestConfig.custom().setConnectTimeout(2000, TimeUnit.MILLISECONDS).build());
            return new URI(address);
        } catch (Exception e) {
            log.error("FlinkCluster getRemoteURI error", e);
        }
        return null;
    }

    @JsonIgnore
    public Map<String, String> getFlinkConfig() throws JsonProcessingException {
        String restUrl = this.address + "/jobmanager/config";
        String json = HttpClientUtils.httpGetRequest(
            restUrl, RequestConfig.custom().setConnectTimeout(2000, TimeUnit.MILLISECONDS).build());
        if (StringUtils.isBlank(json)) {
            return new HashMap<>();
        }
        List<Map<String, String>> confList = JacksonUtils.read(json, new TypeReference<List<Map<String, String>>>() {
        });
        Map<String, String> config = new HashMap<>(0);
        confList.forEach(k -> config.put(k.get("key"), k.get("value")));
        return config;
    }

    @JsonIgnore
    public Map<String, Object> getProperties() {
        Map<String, Object> propertyMap = new HashMap<>();
        Map<String, String> dynamicPropertyMap = PropertiesUtils
            .extractDynamicPropertiesAsJava(this.getDynamicProperties());
        propertyMap.putAll(this.getOptionMap());
        propertyMap.putAll(dynamicPropertyMap);
        ResolveOrder resolveOrder = ResolveOrder.of(this.getResolveOrder());
        if (resolveOrder != null) {
            propertyMap.put(CoreOptions.CLASSLOADER_RESOLVE_ORDER.key(), resolveOrder.getName());
        }
        return propertyMap;
    }

}

File: streampark-console/streampark-console-webapp/src/locales/lang/zh-CN/setting/flinkCluster.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
export default {
  title: 'Flink 集群',
  detail: '查看集群详情',
  stop: '停止集群',
  start: '开启集群',
  edit: '编辑集群',
  delete: '确定要删除此集群 ?',
  searchByName: '根据 Flink 集群名称搜索',
  form: {
    clusterName: '集群名称',
    address: '集群URL',
    runState: '运行状态',
    deployMode: '部署模式',
    versionId: 'Flink版本',
    addType: '添加类型',
    addExisting: '已有集群',
    addNew: '全新集群',
    yarnQueue: 'Yarn队列',
    yarnSessionClusterId: 'Yarn Session模式集群ID',
    k8sNamespace: 'Kubernetes 命名空间',
    k8sClusterId: 'Kubernetes 集群 ID',
    serviceAccount: 'Kubernetes  服务账号',
    k8sConf: 'Kube 配置文件',
    flinkImage: 'Flink 基础 Docker 镜像',
    k8sRestExposedType: 'Kubernetes Rest exposed-type',
    resolveOrder: '类加载顺序',
    taskSlots: '任务槽数',
    jmOptions: 'JM内存',
    tmOptions: 'TM内存',
    dynamicProperties: '动态参数',
    clusterDescription: '集群描述',
  },
  placeholder: {
    addType: '请选择集群添加类型',
    clusterName: '请输入集群名称',
    deployMode: '请选择部署模式',
    versionId: '请选择Flink版本',
    addressRemoteMode: '请输入Flink 集群JobManager URL访问地址',
    addressNoRemoteMode: '请输入集群地址，例如：http://host:port',
    yarnSessionClusterId: '请输入Yarn Session模式集群ID',
    k8sConf: '示例：~/.kube/config',
    flinkImage: '请输入 Flink 基础 docker 镜像的标签，如：flink:1.13.0-scala_2.11-java8',
    k8sRestExposedType: 'kubernetes.rest-service.exposed.type',
    resolveOrder: 'classloader.resolve-order',
    taskSlots: '每个TaskManager的插槽数',
    totalOptions: '总内存',
    jmOptions: '请选择要设置的jm资源参数',
    tmOptions: '请选择要设置的tm资源参数',
    clusterDescription: '集群描述信息, 如: 生产flink 1.16集群',
  },
  required: {
    address: '必须填写集群地址',
    deployMode: '部署模式必填',
    clusterId: 'Yarn Session Cluster 为必填项',
    versionId: 'Flink 版本必选',
    flinkImage: 'link基础docker镜像是必填的',
    resolveOrder: '类加载顺序必选',
  },
  operateMessage: {
    createFlinkSessionClusterSuccessful: ' 创建成功!',
    createFlinkSessionClusterFailed: 'session集群创建失败, 请检查日志',
    hadoopEnvInitializationFailed: 'Hadoop环境初始化失败，请检查环境设置',
    flinkClusterIsStarting: '当前集群正在启动',
    flinkClusterHasStartedSuccessful: '当前集群已成功启动',
    updateFlinkClusterSuccessful: ' 更新成功!',
  },
  view: {
    clusterId: '集群ID',
  },
};
